New conf.py directives autodoc_exclude_members + autodoc_special_members

peterjc opened the pull request:
### Feature or Bugfix
- Feature

### Purpose
- Adds ``conf.py`` setting ``autodoc_exclude_members`` (*update*) and ``autodoc_special_members``

### Relates
- #4057 and #4075 for trying to use ``:exclude-members:`` with apidoc and autodoc.

peterjc committed:
conf.py directive autodoc_exclude_members

peterjc commented:
Note if this change is accepted, I would be happy to try to do the same for ``:special-members:`` as well.

tk0miya commented:
@shimizukawa could you give us advice for this interface?
I don't know what is best design of autodoc.
I also feel to enhance `autodoc_default_flags` to allow dict as a default settings.
what do you think?

stephenfin commented on sphinx/ext/autodoc/__init__.py:
```diff
@@ -752,9 +752,11 @@ def document_members(self, all_members=False):
 
         # remove members given by exclude-members
-        if self.options.exclude_members:
+        exclude = self.options.exclude_members or \
```
nit: I'd rather see this wrapped in brackets as it reads a little better. However, we use `/` elsewhere in the code base so it's not a pressing issue

stephenfin reviewed (approved):
Yup, this is a good idea. Would like to see a similar option it for `special-members` too, as you suggest

peterjc committed:
Wrap with brackets rather than using line continuation slash

Feedback from Stephen Finucane reviewing pull request
#4076.

peterjc committed:
conf.y directive autodoc_special_members

peterjc commented:
This may not be the most elegant way to do this (since I turn the list of special methods into a comma-separated string), but it seems to work OK in my initial testing.

TODO: Testing overrides with different entries in the RST file directives...

peterjc commented:
Sorry for the delay, I've been ill which delayed looking at this.

I may be doing something wrong with my local testing, but it seems neither ``:no-special-members:`` nor entries like ``:special-members: __init__, __eq__`` in an individual RST file are able to override a ``conf.py`` setting like ``autodoc_special_members = ['__init__', '__next__']``

i.e. The special methods part of this patch needs more work. I'm happy to remove this in the short term if you'd rather merge the exclude members work (which I am happy with).

stephenfin commented:
@peterjc If you could point to the exact sections that are causing issues, I'd be happy to take a look this evening?

peterjc commented:
@stephenfin do you know how where the RST file specific autodoc directive for special methods is loaded and applied?

<details>
 <summary>Update: I've started adding logging</summary>

e.g.

```
$ git diff sphinx/ext/autodoc/__init__.py
diff --git a/sphinx/ext/autodoc/__init__.py b/sphinx/ext/autodoc/__init__.py
index 05dea49..064c54b 100644
--- a/sphinx/ext/autodoc/__init__.py
+++ b/sphinx/ext/autodoc/__init__.py
@@ -699,6 +699,12 @@ class Documenter(object):
             keep = False
             if want_all and membername.startswith('__') and \
                     membername.endswith('__') and len(membername) > 4:
+                logger.debug('[autodoc] considering %s.%s with '
+                             'self.options.special_members = %r '
+                             'and self.env.config.autodoc_special_members = %r',
+                             namespace, membername,
+                             self.options.special_members,
+                             self.env.config.autodoc_special_members)
                 # special __methods__
                 if self.options.special_members is ALL and \
                         membername != '__doc__':
@@ -1554,6 +1560,10 @@ class AutoDirective(Directive):
                 self.options[flag] = None
 
         if self.env.config.autodoc_special_members:
+            logger.debug("[autodoc] applying autodoc_special_members = %r, "
+                         "currently self.options['special-members'] = %r",
+                         self.env.config.autodoc_special_members,
+                         self.options['special-members'])
             # Using autodoc_special_members (which takes a list of methods)
             # overrides including special-members in autodoc_default_flags
             # (which applies the default value of None meaning ALL methods).
```

This output running ``sphinx-build -v ...`` to see the logging puzzles me:

```
...
[autodoc] /mnt/shared/users/xxx/repositories/biopython/Doc/api/Bio.SeqRecord.rst:4: input:
.. automodule:: Bio.SeqRecord
    :members:
    :undoc-members:
    :show-inheritance:
    :special-members: __init__,__eq__
[autodoc] applying autodoc_special_members = ['__init__', '__next__'], currently self.options['special-members'] = None
[autodoc] import Bio.SeqRecord
...
```

At what point does this RST file's entry ``:special-members: __init__,__eq__`` get loaded?

Note my ``conf.py`` file contains ``autodoc_special_members = ['__init__', '__next__']`` in this example.
</details>

peterjc commented:
To try to understand what is breaking,I'm wondering what the documentation update https://github.com/sphinx-doc/sphinx/commit/c333f515cd9a3b020c3bf3b31db89262e9143536 refers to - this added the text for ``:special-members:`` saying as of release 1.2 *"The option can now take arguments, i.e. the special members to document."*

The git commit comment is *"Doc change for d7ac5e46307d"* but that hash does not seem to be in the tree...

```
$ git show d7ac5e46307d
fatal: ambiguous argument 'd7ac5e46307d': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

*Update:* I'm guessing it refers to https://github.com/sphinx-doc/sphinx/commit/5b9f431fc92fc76c01fcb3b86c02c37298f1c2fc *"Changed the ``special-members`` option to behave like the ``members`` option."* - which is all bits of the code I was already looking at. Hmm.

peterjc committed:
Fix autodoc_special_members priority.

Entry in the RST file most important, then conf.py entry
autodoc_special_members (list), finally conf.py entry
autodoc_default_flags (where inclusion here means ALL
special methods).

peterjc commented:
As part of fixing the priorities for the new ``autodoc_special_members`` setting, I thinks I also have found and fixed a pre-existing bug where ``autodoc_default_flags = ['special-members', ...]`` would override any RST file entry like ``:special-members: __init__,__eq__`` or ``:no-special-members:``

peterjc committed:
Reduce line length for flake8

peterjc commented:
Is there anything else that I can improve/fix before this is reviewed by @shimizukawa?

shimizukawa commented:
> I also feel to enhance autodoc_default_flags to allow dict as a default settings.

+1 (by a quick look..)

adamklein commented:
+1

tk0miya commented:
Now I merged #5215 instead.
Thank you for your work!

peterjc commented:
That's great - I'll give this new functionality a test next time I work on our documentation. Thank you 😀

peterjc commented:
That's great - I'll give this new functionality a test next time I work on our documentation. Thank you 😀
